draft-handrews-jri.md

```diff
@@ -0,0 +1,536 @@
+
+These parameters are intended for use in environments where a standalone implementation is likely to encounter documents with unexpected media types involving unpredictable JRI usage.  Standalone implementations MAY use other heuristics and information sources to determine whether to process a document.
+
+Non-media type context data formats MAY define a mechanism analogous to these media type parameters, but such mechanisms will not be interoperable.
+
+#### The "jri-safety" media type parameter
+
+The value of the `jri-safety` parameter MUST be case-insensitive, and MUST be either `interoperable` or `contextual`.  The default value is `contextual`, indicating that an implementation MUST be aware of and support the relevant context specification in order to safely process JRI keywords in the document.  A value of `interoperable` indicates that the document is context-independent, and safe to process by a standalone implementation.
+
+### The "jri-keywords" media type parameter
+
+The value of the `jri-keywords` parameter MUST be a case-sensitive comma-separated list of JRI keyword names, which MUST NOT contain whitespace.  These keywords MAY be in any order.  This indicates that only the listed keywords are used in the document. In the absence of this parameter, an implementation MUST assume that any JRI keyword can be used.
+
+### A JSON Schema vocabulary for safe processing
+
+[^42]
+
+[^42]: A simple annotation vocabulary could indicate where exactly in a document JRI keywords can appear, and what behavior they might have.  Unlike the media type parameters, this approach could distinguish between `"$ref"`/`"$id"`/`"$anchor"` within a schema vs those property names with JRI-correct values within `enum`, `const`, or `examples`.  This would make it possible to use a generic JRI cache as the base for a JSON Schema implementation.
+
+# Security Considerations
+
+On-demand retrieval and processing of referenced resources brings the security risks inherent in handling any data from an untrusted source.  It is RECOMMENDED that implementations supporting this behavior support defining allow/deny lists for limiting retrieval to trusted sources.
+
+Implementations of a JRI cache that implement a functional caching system (rather than a simple lookup table) need to address the security considerations of the caching process(es) that they support.
+
+## Guarding Against Infinite Recursion
+
+An implementation MUST guard against infinite reference loops.
+
+Processing rules for context specifications MAY make it safe to have cyclic references by creating conditions under which a reference will or will not be followed.
+
+Cyclic references cannot safely be processed in a way that is both interoperable and context-independent.  Seeing a specific reference while processing the target of that same reference MUST be considered an error by context-unaware implementations.
```

Comment:
can you clarify this, especially the part about  "Seeing a specific reference" -- does that mean before or  after resolving the base URI, etc.? 
For example in our OAS documents, we define a schema for an [RFC 7807](https://www.rfc-editor.org/rfc/rfc7807#section-3.1) problem object (called `apiProblem`) and extended it with a `problems` array of `apiProblem` objects, in this case defined with an array whose `items` are`$ref: '#/components/schemas/apiProblem'`  Does that violate this clause and MUST be considered an error?
We also extract (debundle) these schemas into context-independent files.

Reply:
I think this was a casualty of me reorganizing sections.  It's an error if you are trying to remove references, but in other situations it's completely fine as you note.  I'll rework this.